internal/credential/credential_libraries.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package credential
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/util"
+)
+
+// LibraryRepository defines the interface expected
+// to get the total number of credential libraries and deleted ids.
+type LibraryRepository interface {
+	GetTotalLibraries(context.Context) (int, error)
+	GetTotalSSHCertificateLibraries(context.Context) (int, error)
+	ListDeletedLibraryIds(context.Context, time.Time, ...Option) ([]string, error)
+	ListDeletedSSHCertificateLibraryIds(context.Context, time.Time, ...Option) ([]string, error)
+}
```

Comment:
This interface seems a bit odd. For one, it is not clear to me that this collection of methods constitutes a "repository". It might be the case that the vault.Repository implements this interface, but that doesn't mean this combination of methods *is* a repository.

Also, does `GetTotalLibraries` include the count of `GetTotalSSHCertificateLibraries`? The names seem to imply that it would, but then why have both?

And lastly, while we currently only have credential libraries from the vault package. In the future we might have additional ones. I would imagine that other credential libraries would have types that do not overlap with vault, so it seems odd that they would need to implement a "GetTotalSSHCertificateLibraries" method if only vault provides SSHCertificateLibraries. Perhaps it would help to create a service in the vault package that can aggregate all vault credential libraries. Then the service in the credential package can aggregate all of the libraries from those services (which would currently just be the vault one).

Reply:
That's fair, I think we can more accurately name this `VaultLibraryRepository`, which at least helps make it clear that this repository is not necessarily expected to be implemented by other credential library repositories. If we add other credential library types in the future they would likely get their own interface definition. That way we wouldn't need to add a service wrapping all Vault library interactions.

The naming situation comes from following the naming conventions. In the Vault repository, the separation between "libraries" (sometimes called "generic libraries") and "ssh certificate libraries" is by file and by name. In implementing methods that operate on the "generic library" type, we're following the convention of just calling it "library", in the context of the Vault repository. Same thing with calling it "ssh certificate library" for SSH certificate libraries. I agree that this repository definition looks strange as a result, but the alternative I see is to rename the use of the "library" term to be "generic library" in the Vault repository. While `GetTotalLibraries` exists, it does not include the count for all library types in the Vault repository. I'd rather not make a rename from "Library" to "GenericLibrary" in the Vault repository in this PR, but I _do_ agree that such a change would be beneficial.

Reply:
I've renamed this interface, do you think that is enough?